### PR TITLE
Refactor context to track self/local types

### PR DESCRIPTION
**Pull request by maximecb**

### CODE REVIEW

File: yjit_core.c [hunk 1]
```diff
@@ -90,18 +113,58 @@ ctx_stack_opnd(ctx_t* ctx, int32_t idx)
...
+*/
+int type_diff(val_type_t src, val_type_t dst)
+{
+    RUBY_ASSERT(!src.is_heap || !src.is_imm);
+    RUBY_ASSERT(!dst.is_heap || !dst.is_imm);
+
+    if (src.type != dst.type && dst.type != ETYPE_UNKNOWN)
+        return INT_MAX;
+
+    if (src.is_heap && !dst.is_heap)
```
Comment (XrXr) [inline yjit_core.c]: Shouldn't this be `!src.is_heap && dst.is_heap`? `is_heap == 0` means `dst` doesn't make any assumptions about it being heap no?

**Comment by XrXr** [comment]
Not sure what's going on with the failing test. I'll have to take a closer look later.

File: yjit_core.c (hunk 1 above)
Comment (maximecb) [reply yjit_core.c]: Good catch. Fixing it :)

**Comment by maximecb** [comment]
Looks like YJIT first call failed on another PR as well, so thankfully it's not directly related to these changes.

**Pull request merged**